File: mobile/android/app/src/main/java/com/radiostream/player/PlaylistPlayer.java
package com.radiostream.player;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;
import com.radiostream.networking.metadata.MetadataBackendGetter;
import com.radiostream.util.SetTimeout;

import org.jdeferred.DoneCallback;
import org.jdeferred.DonePipe;
import org.jdeferred.FailCallback;
import org.jdeferred.Promise;
import org.jdeferred.impl.DeferredObject;

import javax.inject.Inject;

import timber.log.Timber;

public class PlaylistPlayer implements Song.EventsListener, PlaylistControls {
    private final StatusProvider mStatusProvider;
    private SetTimeout mSetTimeout;
    private Playlist mPlaylist;
    private Song mCurrentSong;
    private MetadataBackendGetter mMetadataBackendGetter;

    private boolean mIsLoading = false;
    private Exception mLastLoadingError = null;

    private boolean mIsClosed = false;


    @Inject
    public PlaylistPlayer(Playlist playlist, SetTimeout setTimeout,
                          MetadataBackendGetter metadataBackendGetter, StatusProvider statusProvider) {

        mPlaylist = playlist;
        mStatusProvider = statusProvider;
        mSetTimeout = setTimeout;
        mMetadataBackendGetter = metadataBackendGetter;
    }

    private void setSongLoadingStatus(boolean isLoading, Exception error) {
        Timber.i("change loading to: %b and error to: %s", isLoading, error != null ? error.toString() : "NULL");
        if (isLoading != mIsLoading || mLastLoadingError != error) {
            Timber.i("value changed");
            mIsLoading = isLoading;
            mLastLoadingError = error;
            mStatusProvider.sendStatus();
        } else {
            Timber.i("value didn't change");
        }
    }

    public Song getCurrentSong() {
        return mCurrentSong;
    }

    private void setCurrentSong(Song value) {
        if (value != getCurrentSong()) {
            if (getCurrentSong() != null) {
                getCurrentSong().pause();
                getCurrentSong().close();
            }

            Timber.i("changing current song to: %s", value.toString());
            mCurrentSong = value;
            mStatusProvider.sendStatus();
        }
    }

    @Override
    public Promise<Song, Exception, Void> play() {
        Timber.i("function start");
        if (mIsLoading) {
            Timber.i("invalid request. song already loading");
            throw new IllegalStateException("invalid request. song already loading");
        }

        Promise<Song, Exception, Void> promise;

        if (getCurrentSong() != null && mPlaylist.isCurrentSong(getCurrentSong())) {
            Timber.i("playing paused song");
            getCurrentSong().subscribeToEvents(PlaylistPlayer.this);
            getCurrentSong().play();

            mStatusProvider.sendStatus();

            promise = new DeferredObject<Song, Exception, Void>().resolve(getCurrentSong()).promise();
        } else {
            Timber.i("loading different song from playlist");
            promise = retryPreloadAndPlaySong()
                .then(new DonePipe<Song, Song, Exception, Void>() {
                    @Override
                    public Promise<Song, Exception, Void> pipeDone(Song result) {
                        return PlaylistPlayer.this.preloadPeekedSong();
                    }
                });
        }

        return promise;
    }

    @Override
    public void pause() {
        Timber.i("function start");

        if (mIsLoading || getCurrentSong() == null) {
            throw new IllegalStateException("no song was loaded yet");
        }

        getCurrentSong().pause();
        mStatusProvider.sendStatus();
    }

    public boolean getIsPlaying() {
        if (getCurrentSong() == null) {
            return false;
        } else {
            return getCurrentSong().isPlaying();
        }
    }

    @Override
    public void playNext() {
        Timber.i("function start");
        this.mPlaylist.nextSong();
        this.play();
    }

    private Promise<Song, Exception, Void> retryPreloadAndPlaySong() {
        Timber.i("function start");

        // Due to: https://github.com/jdeferred/jdeferred/issues/20
        // To convert a failed promise to a resolved one, we must create a new deferred object
        final DeferredObject<Song, Exception, Void> deferredObject = new DeferredObject<>();

        // NOTE: the last error to the function is sent since merely retrying
        // the function doesn't clear the erro
        setSongLoadingStatus(true, mLastLoadingError);
        mStatusProvider.sendStatus();

        waitForCurrentSongMarkedAsPlayed().then(new DonePipe<Boolean, Song, Exception, Void>() {
            @Override
            public Promise<Song, Exception, Void> pipeDone(Boolean result) {
                return mPlaylist.peekCurrentSong();
            }
        }).then(new DonePipe<Song, Song, Exception, Void>() {
            @Override
            public Promise<Song, Exception, Void> pipeDone(Song song) {
                try {
                    Timber.i("preloading song: %s", song.toString());
                    setCurrentSong(song);
                    return song.preload();
                } catch (Exception e) {
                    return new DeferredObject<Song, Exception, Void>().reject(e).promise();
                }
            }
        }).then(new DonePipe<Song, Song, Exception, Void>() {
            @Override
            public Promise<Song, Exception, Void> pipeDone(Song song) {
                try {
                    setSongLoadingStatus(false, null);

                    // We won't be playing any new music if playlistPlayer is closed
                    if (!mIsClosed) {
                        PlaylistPlayer.this.play();
                    } else {
                        Timber.i("playlist player was already closed - not playing loaded song");
                    }

                    deferredObject.resolve(song);

                    return deferredObject.promise();
                } catch (Exception e) {
                    return new DeferredObject<Song, Exception, Void>().reject(e).promise();
                }
            }
        }).fail(new FailCallback<Exception>() {
            @Override
            public void onFail(Exception exception) {
                Timber.e(exception, "exception occured during next song loading");
                setSongLoadingStatus(true, exception);
                deferredObject.resolve(null);
            }
        });

        return deferredObject.promise().then(new DonePipe<Song, Song, Exception, Void>() {
            @Override
            public Promise<Song, Exception, Void> pipeDone(Song song) {
                if (song == null) {
                    Timber.i("no song was loaded - waiting and retrying");
                    return mSetTimeout.run(10000).then(new DonePipe<Void, Song, Exception, Void>() {
                        @Override
                        public Promise<Song, Exception, Void> pipeDone(Void result) {
                            Timber.i("timeout finished - retrying");
                            return PlaylistPlayer.this.retryPreloadAndPlaySong();
                        }
                    });
                } else {
                    Timber.i("song preloaded successfully");
                    return new DeferredObject<Song, Exception, Void>().resolve(song).promise();
                }
            }
        });
    }

    private Promise<Boolean, Exception, Void> waitForCurrentSongMarkedAsPlayed() {
        Timber.i("function start");
        if (getCurrentSong() != null) {
            return getCurrentSong().waitForMarkedAsPlayed();
        } else {
            Timber.i("no current song found - not waiting");
            return new DeferredObject<Boolean, Exception, Void>().resolve(true).promise();
        }
    }

    private Promise<Song, Exception, Void> preloadPeekedSong() {
        Timber.i("function start");

        return mPlaylist.peekNextSong()
            .then(new DonePipe<Song, Song, Exception, Void>() {
                @Override
                public Promise<Song, Exception, Void> pipeDone(Song peekedSong) {
                    Timber.i("preloading peeked song: %s", peekedSong.toString());
                    return peekedSong.preload();
                }
            }).fail(new FailCallback<Exception>() {
                @Override
                public void onFail(Exception error) {
                    Timber.w("failed to preload song: %s", error.toString());
                }
            });
    }

    void close() {
        Timber.i("function start");
        mIsClosed = true;

        if (getCurrentSong() != null) {
            Timber.i("closing current song");
            getCurrentSong().close();
        }

        if (mPlaylist != null) {
            mPlaylist.close();
            mPlaylist = null;
        }
    }

    @Override
    public void onSongFinish(Song song) {
        Timber.i("function start");
        this.playNext();
    }

    @Override
    public void onSongMarkedAsPlayed() {
        mStatusProvider.sendStatus();
    }

    @Override
    public void onSongError(Exception error) {
        Timber.e(error, "error occured in song '%s'", getCurrentSong());
        if (getCurrentSong() != null) {
            Timber.i("pausing existing song");
            getCurrentSong().pause();
        }

        Timber.i("trying to play next song");
        play();
    }

    public WritableMap toBridgeObject() {
        WritableMap map = Arguments.createMap();
        map.putBoolean("isLoading", mIsLoading);
        map.putBoolean("isPlaying", getIsPlaying());
        map.putMap("playlist", mPlaylist != null ? mPlaylist.toBridgeObject() : null);
        map.putString("loadingError", mLastLoadingError != null ? mLastLoadingError.getMessage() : "");

        return map;
    }

    public Promise<Void, Exception, Void> updateSongRating(int songId, final int newRating) {
        Timber.i("function start");
        final Song updatedSong = getCurrentSong();
        if (updatedSong.getId() == songId) {
            return mMetadataBackendGetter.get().updateSongRating(songId, newRating).then(new DoneCallback<Void>() {
                @Override
                public void onDone(Void result) {
                    updatedSong.setRating(newRating);

                    Timber.i("song rating updated - sending status update");
                    mStatusProvider.sendStatus();
                }
            });
        } else {
            Timber.w("tried to update id %d even though current song %s has id %d",
                songId, updatedSong.toString(), updatedSong.getId());

            return new DeferredObject<Void, Exception, Void>().reject(null);
        }
    }
}

File: app/android/app/src/main/java/com/radiostream/player/PlaylistPlayer.kt
package com.radiostream.player

import com.facebook.react.bridge.WritableMap
import com.radiostream.wrapper.ArgumentsInterface
import com.radiostream.networking.metadata.MetadataBackendGetter
import com.radiostream.ui.PlayerNotification
import kotlinx.coroutines.experimental.*

import javax.inject.Inject

import timber.log.Timber

class PlaylistPlayer @Inject
constructor(private var mPlaylist: Playlist?, private val mMetadataBackendGetter: MetadataBackendGetter, private val mStatusProvider: StatusProvider,
            private val mArguments: ArgumentsInterface) : Song.EventsListener, PlaylistControls {

    var currentSong: Song? = null
        private set(value) {
            field = value
        }

    var isLoading = false
        private set(value) {
            field = value
        }

    private var mLastLoadingErrorMessage: String = ""

    private var mIsClosed = false

    val isPlaying: Boolean
        get() = if (currentSong == null) {
            false
        } else {
            currentSong!!.isPlaying
        }

    private fun setSongLoadingStatus(isLoading: Boolean, errorMessage: String = "") {
        Timber.i("change loading to: %b and error to: %s", isLoading, errorMessage)
        if (isLoading != this.isLoading || mLastLoadingErrorMessage !== errorMessage) {
            Timber.i("value changed")
            this.isLoading = isLoading
            mLastLoadingErrorMessage = errorMessage
            mStatusProvider.sendStatus()
        } else {
            Timber.i("value didn't change")
        }
    }

    override suspend fun play() {
        Timber.i("function start")
        if (isLoading) {
            Timber.i("invalid request. song already loading")
            throw IllegalStateException("invalid request. song already loading")
        }

        if (currentSong != null && mPlaylist!!.isCurrentSong(currentSong!!)) {
            Timber.i("playing paused song")
            currentSong!!.subscribeToEvents(this@PlaylistPlayer)
            currentSong!!.play()

            mStatusProvider.sendStatus()
        } else {
            Timber.i("loading different song from playlist")
            retryPreloadAndPlaySong()
            preloadPeekedSong()
        }
    }

    override fun pause() {
        Timber.i("function start")

        if (isLoading || currentSong == null) {
            throw IllegalStateException("no song was loaded yet")
        }

        currentSong!!.pause()
        mStatusProvider.sendStatus()
    }

    override suspend fun playNext() {
        Timber.i("function start")
        this.mPlaylist!!.nextSong()
        this.play()
    }

    override suspend fun skipToSongByIndex(index: Int) {
        Timber.i("function start")

        this.mPlaylist!!.skipToIndex(index)
        this.play()
    }

    private suspend fun retryPreloadAndPlaySong() {
        Timber.i("function start")

        try {
            setSongLoadingStatus(true, mLastLoadingErrorMessage)
            if (currentSong != null) {
                currentSong!!.pause()
                currentSong!!.close()

                currentSong = null
            }

            mStatusProvider.sendStatus()

            waitForCurrentSongMarkedAsPlayed()
            val peekedSong = mPlaylist!!.peekCurrentSong()
            Timber.i("peeked song: $peekedSong")
            Timber.i("preloading song: %s", peekedSong.toString())

            mStatusProvider.sendStatus()
            peekedSong.preload()
            setSongLoadingStatus(false)

            // We won't be playing any new music if playlistPlayer is closed
            if (!mIsClosed) {
                Timber.i("changing current song to: %s", peekedSong.toString())
                currentSong = peekedSong
                mStatusProvider.sendStatus()
                play()
            } else {
                Timber.i("playlist player was already closed - not playing loaded song")
            }

            Timber.i("song preloaded successfully")
        } catch (e: Exception) {
            Timber.e(e, "exception occurred during next song loading")
            setSongLoadingStatus(true, "Failed. Retrying...")

            Timber.i("no song was loaded - waiting and retrying")
            delay(5000)
            Timber.i("timeout finished - retrying")
            retryPreloadAndPlaySong()
        }
    }

    private suspend fun waitForCurrentSongMarkedAsPlayed() {
        Timber.i("function start")
        if (currentSong != null) {
            currentSong!!.waitForMarkedAsPlayed()
        } else {
            Timber.i("no current song found - not waiting")
        }
    }

    private suspend fun preloadPeekedSong() {
        Timber.i("function start")

        try {
            val peekedSong = mPlaylist!!.peekNextSong()
            Timber.i("preloading peeked song: %s", peekedSong.toString())
            peekedSong.preload()
        } catch (e: Exception) {
            Timber.w("failed to preload song: %s", e.toString())
        }
    }

    fun close() {
        Timber.i("function start")
        mIsClosed = true

        if (currentSong != null) {
            Timber.i("closing current song")
            currentSong!!.close()
        }

        if (mPlaylist != null) {
            mPlaylist!!.close()
            mPlaylist = null
        }
    }

    override suspend fun onSongFinish(song: Song) {
        Timber.i("function start")
        this.playNext()
    }

    override fun onSongMarkedAsPlayed() {
        mStatusProvider.sendStatus()
    }

    override suspend fun onSongError(error: Exception) {
        Timber.e(error, "error occurred in song '%s'", currentSong)
        if (currentSong != null) {
            Timber.i("pausing existing song")
            currentSong!!.pause()
        }

        Timber.i("trying to play next song")
        playNext()
    }

    fun toBridgeObject(): WritableMap {
        val map = mArguments.createMap()
        map.putBoolean("isLoading", isLoading)
        map.putBoolean("isPlaying", isPlaying)
        map.putMap("playlist", if (mPlaylist != null) mPlaylist!!.toBridgeObject() else null)
        map.putString("loadingError", mLastLoadingErrorMessage)

        return map
    }

    suspend fun updateSongRating(songId: Int, newRating: Int) {
        Timber.i("function start")
        val updatedSong = currentSong
        if (updatedSong!!.id == songId) {
            mMetadataBackendGetter.get().updateSongRating(songId, newRating)
            updatedSong.setRating(newRating)
            Timber.i("song rating updated - sending status update")
            mStatusProvider.sendStatus()
        } else {
            Timber.w("tried to update id %d even though current song %s has id %d",
                    songId, updatedSong.toString(), updatedSong.id)
        }
    }
}
